WalletWasabi/WabiSabi/Models/MultipartyTransaction/Signing.cs

```diff
@@ -0,0 +1,88 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using WalletWasabi.WabiSabi.Backend.Models;
+
+namespace WalletWasabi.WabiSabi.Models.MultipartyTransaction
+{
+	public record Signing(Parameters Parameters, ImmutableArray<Coin> Inputs, ImmutableArray<TxOut> Outputs) : State
+	{
+		public ImmutableDictionary<int, WitScript> Witnesses { get; init; } = ImmutableDictionary<int, WitScript>.Empty;
+
+		public bool IsFullySigned => Witnesses.Count() == Inputs.Length;
+
+		public FeeRate EffectiveFeeRate => new FeeRate(Balance, EstimatedVsize);
+
+		public Money Balance => Inputs.Sum(x => x.Amount) - Outputs.Sum(x => x.Value);
+
+		public int EstimatedVsize => Parameters.SharedOverhead + EstimatedInputsVsize + OutputsVsize;
+		public int EstimatedInputsVsize => Inputs.Sum(x => x.TxOut.ScriptPubKey.EstimateInputVsize());
+		public int OutputsVsize => Outputs.Sum(x => x.ScriptPubKey.EstimateOutputVsize());
+
+		public IEnumerable<Coin> UnsignedInputs => Enumerable.Range(0, Inputs.Length).Where(i => !IsInputSigned(i)).Select(i => Inputs[i]);
+
+		public bool IsInputSigned(int index) => Witnesses.ContainsKey(index);
+
+
+		public Signing AddWitness(int index, WitScript witness)
+		{
+			if (IsInputSigned(index))
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WitnessAlreadyProvided);
+			}
+
+			// Verify witness.
+			// 1. Copy UnsignedCoinJoin.
+			Transaction cjCopy = CreateUnsignedTransaction();
+
+			// 2. Sign the copy.
+			cjCopy.Inputs[index].WitScript = witness;
+
+			// 3. Convert the current input to IndexedTxIn.
+			IndexedTxIn currentIndexedInput = cjCopy.Inputs.AsIndexedInputs().Skip(index).First();
+
+			// 4. Find the corresponding registered input.
+			Coin registeredCoin = Inputs[index];
+
+			// 5. Verify if currentIndexedInput is correctly signed, if not, return the specific error.
+			if (!currentIndexedInput.VerifyScript(registeredCoin, out ScriptError error))
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongCoinjoinSignature); // TODO keep script error
+			}
+
+			return this with { Witnesses = Witnesses.Add(index, witness) };
+		}
+
+		public Transaction CreateUnsignedTransaction()
+		{
+			var tx = Parameters.CreateTransaction();
+
+			foreach (var coin in Inputs)
+			{
+				// implied:
+				// nSequence = FINAL
+				tx.Inputs.Add(coin.Outpoint);
+			}
+
+			foreach (var txout in Outputs)
+			{
+				tx.Outputs.AddWithOptimize(txout.Value, txout.ScriptPubKey);
+			}
+
```

Comment:
Untested. 

```c#
tx.Inputs.AddRange(Inputs.Select(x => new TxIn(x.Outpoint)));
tx.Outputs.AddRange(Outputs.GroupBy(x => x.ScriptPubKey, x => x.Value).Select(x => new TxOut(x.Sum(), x.Key)));
```

I am sure the outputs consolidation can be improved but I don't know if you agree with the syntax.

Reply:
I think for the 1st it's definitely the way to go.

for the second, this effectively reimplements `AddWithOptimize` so I wasn't sure if that's desirable, that's why I didn't apply yet